apps/explorer/lib/explorer/chain/transaction/state_change.ex

```diff
@@ -4,50 +4,94 @@ defmodule Explorer.Chain.Transaction.StateChange do
+          address: Address.t(),
           token_id: nil | non_neg_integer(),
           balance_before: Wei.t() | Decimal.t(),
           balance_after: Wei.t() | Decimal.t(),
           balance_diff: Wei.t() | Decimal.t(),
           miner?: boolean()
         }
 
-  def coin_balances_before(tx, block_txs, from_before, to_before, miner_before) do
+  @type coin_balances_map :: %{Hash.Address.t() => {Address.t(), Wei.t()}}
+
+  @zero_wei %Wei{value: Decimal.new(0)}
+
+  @spec coin_balances_before(Transaction.t(), [Transaction.t()], coin_balances_map()) :: coin_balances_map()
+  def coin_balances_before(tx, block_txs, coin_balances_before_block) do
     block = tx.block
 
     block_txs
     |> Enum.reduce_while(
-      {from_before, to_before, miner_before},
-      fn block_tx, {block_from, block_to, block_miner} = state ->
+      coin_balances_before_block,
+      fn block_tx, acc ->
         if block_tx.index < tx.index do
-          {:cont,
-           {update_coin_balance_from_tx(tx.from_address_hash, block_tx, block_from, block),
-            update_coin_balance_from_tx(tx.to_address_hash, block_tx, block_to, block),
-            update_coin_balance_from_tx(tx.block.miner_hash, block_tx, block_miner, block)}}
+          {:cont, update_coin_balances_from_tx(acc, block_tx, block)}
         else
           # txs ordered by index ascending, so we can halt after facing index greater or equal than index of our tx
-          {:halt, state}
+          {:halt, acc}
         end
       end
     )
   end
 
-  def update_coin_balance_from_tx(address_hash, tx, balance, block) do
-    from = tx.from_address_hash
-    to = tx.to_address_hash
-    miner = block.miner_hash
+  @spec update_coin_balances_from_tx(coin_balances_map(), Transaction.t(), Block.t()) :: coin_balances_map()
+  def update_coin_balances_from_tx(coin_balances, tx, block) do
+    coin_balances =
+      coin_balances
+      |> (&if(Map.has_key?(coin_balances, tx.from_address_hash),
+            do:
+              Map.update(&1, tx.from_address_hash, @zero_wei, fn {address, balance} ->
+                {address, Wei.sub(balance, from_loss(tx))}
+              end),
+            else: &1
+          )).()
+      |> (&if(tx.to_address_hash && Map.has_key?(coin_balances, tx.to_address_hash),
+            do:
+              Map.update(&1, tx.to_address_hash, @zero_wei, fn {address, balance} ->
+                {address, Wei.sum(balance, to_profit(tx))}
+              end),
+            else: &1
+          )).()
+      |> (&if(Map.has_key?(coin_balances, block.miner_hash),
+            do:
+              Map.update(&1, block.miner_hash, @zero_wei, fn {address, balance} ->
+                {address, Wei.sum(balance, miner_profit(tx, block))}
+              end),
+            else: &1
+          )).()
+
+    if error?(tx) do
+      coin_balances
+    else
+      tx.internal_transactions |> Enum.reduce(coin_balances, &update_coin_balances_from_internal_tx(&1, &2))
+    end
+  end
```

Comment:
Don't bother my comment if I am wrong, but I think we could use here regular function instead of multiple anonymous functions

```suggestion
  def update_coin_balances_from_tx(coin_balances, tx, block) do
    coin_balances =
      coin_balances
      |> update_balance(tx.from_address_hash, &Wei.sub(&1, from_loss(tx)))
      |> update_balance(tx.to_address_hash, &Wei.sum(&1, to_profit(tx)))
      |> update_balance(block.miner_hash, &Wei.sum(&1, miner_profit(tx, block)))

    if error?(tx) do
      coin_balances
    else
      tx.internal_transactions |> Enum.reduce(coin_balances, &update_coin_balances_from_internal_tx(&1, &2))
    end
  end

  defp update_balance(coin_balances, address_hash, update_fn) when is_nil(address_hash), 
    do: coin_balances
  
  defp update_balance(coin_balances, address_hash, update_function) do
    if Map.has_key?(coin_balances, address_hash) do
      Map.update(coin_balances, address_hash, @zero_wei, fn {address, balance} ->
        {address, update_function.(balance)}
      end)
    else
      coin_balances
    end
  end
```

Reply:
Good suggestion, thanks! Refactored here https://github.com/blockscout/blockscout/pull/10210/commits/b95bc189c0f3c31393d27b6dedc4a0f73dfb22b2